maven-di/src/main/java/org/apache/maven/di/ResourceLocator.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.maven.di;
+
+import java.util.function.Supplier;
+
+import org.apache.maven.api.annotations.Nullable;
+
+public interface ResourceLocator {
+    <T> T getInstance(Key<T> key);
+
+    default <T> T getInstance(Class<T> type) {
+        return getInstance(Key.of(type));
+    }
+
+    <T> @Nullable T getInstanceOrNull(Key<T> key);
```

Comment:
guess all the next getInstances can be replaced by getOptionalInstance and the cases be managed by `Optional` directly instead of duplicating API?